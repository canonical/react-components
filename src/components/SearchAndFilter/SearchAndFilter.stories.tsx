import { Meta, StoryObj } from "@storybook/react";

import SearchAndFilter from "./SearchAndFilter";

const meta: Meta<typeof SearchAndFilter> = {
  component: SearchAndFilter,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof SearchAndFilter>;

export const Default: Story = {
  name: "Default",

  args: {
    returnSearchData: () => {},

    filterPanelData: [
      {
        id: 0,
        heading: "Cloud",

        chips: [
          {
            lead: "Cloud",
            value: "Google",
          },
          {
            lead: "Cloud",
            value: "AWS",
          },
          {
            lead: "Cloud",
            value: "Azure",
          },
        ],
      },
      {
        id: 1,
        heading: "Region",

        chips: [
          {
            lead: "Region",
            value: "us-east1",
          },
          {
            lead: "Region",
            value: "us-north2",
          },
          {
            lead: "Region",
            value: "us-south3",
          },
          {
            lead: "Region",
            value: "us-north4",
          },
          {
            lead: "Region",
            value: "us-east5",
          },
          {
            lead: "Region",
            value: "us-south6",
          },
          {
            lead: "Region",
            value: "us-east7",
          },
          {
            lead: "Region",
            value: "us-east8",
          },
          {
            lead: "Region",
            value: "us-east9",
          },
          {
            lead: "Region",
            value: "us-east10",
          },
        ],
      },
      {
        id: 2,
        heading: "Owner",

        chips: [
          {
            lead: "Owner",
            value: "foo",
          },
          {
            lead: "Owner",
            value: "bar",
          },
          {
            lead: "Owner",
            value: "baz",
          },
        ],
      },
    ],
  },
};

export const WithDataSet: Story = {
  name: "With data set",

  args: {
    filterPanelData: [
      {
        id: 0,
        heading: "Cloud",

        chips: [
          {
            lead: "Cloud",
            value: "Google",
          },
          {
            lead: "Cloud",
            value: "AWS",
          },
          {
            lead: "Cloud",
            value: "Azure",
          },
        ],
      },
      {
        id: 1,
        heading: "Region",

        chips: [
          {
            lead: "Region",
            value: "us-east1",
          },
          {
            lead: "Region",
            value: "us-north2",
          },
          {
            lead: "Region",
            value: "us-south3",
          },
          {
            lead: "Region",
            value: "us-north4",
          },
          {
            lead: "Region",
            value: "us-east5",
          },
          {
            lead: "Region",
            value: "us-south6",
          },
          {
            lead: "Region",
            value: "us-east7",
          },
          {
            lead: "Region",
            value: "us-east8",
          },
          {
            lead: "Region",
            value: "us-east9",
          },
          {
            lead: "Region",
            value: "us-east10",
          },
        ],
      },
      {
        id: 2,
        heading: "Owner",

        chips: [
          {
            lead: "Owner",
            value: "foo",
          },
          {
            lead: "Owner",
            value: "bar",
          },
          {
            lead: "Owner",
            value: "baz",
          },
        ],
      },
    ],

    returnSearchData: () => {},
  },
};

export const WithExistingSearchData: Story = {
  name: "With existing search data",

  args: {
    existingSearchData: [
      {
        lead: "Cloud",
        value: "Google",
      },
    ],

    filterPanelData: [
      {
        id: 0,
        heading: "Cloud",

        chips: [
          {
            lead: "Cloud",
            value: "Google",
          },
          {
            lead: "Cloud",
            value: "AWS",
          },
          {
            lead: "Cloud",
            value: "Azure",
          },
        ],
      },
      {
        id: 1,
        heading: "Region",

        chips: [
          {
            lead: "Region",
            value: "us-east1",
          },
          {
            lead: "Region",
            value: "us-north2",
          },
          {
            lead: "Region",
            value: "us-south3",
          },
          {
            lead: "Region",
            value: "us-north4",
          },
          {
            lead: "Region",
            value: "us-east5",
          },
          {
            lead: "Region",
            value: "us-south6",
          },
          {
            lead: "Region",
            value: "us-east7",
          },
          {
            lead: "Region",
            value: "us-east8",
          },
          {
            lead: "Region",
            value: "us-east9",
          },
          {
            lead: "Region",
            value: "us-east10",
          },
        ],
      },
      {
        id: 2,
        heading: "Owner",

        chips: [
          {
            lead: "Owner",
            value: "foo",
          },
          {
            lead: "Owner",
            value: "bar",
          },
          {
            lead: "Owner",
            value: "baz",
          },
        ],
      },
    ],

    returnSearchData: () => {},
  },
};
