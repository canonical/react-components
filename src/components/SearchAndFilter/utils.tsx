/**
 * Return number of overflowing chips given a row threshold
 * @param {array} chips - An array of chips
 * @param {Integer} overflowRowLimit - Number of rows to show before counting
 * overflow
 */
export const overflowingChipsCount = (chips, overflowRowLimit) => {
  let overflowChips = 0;
  if (chips) {
    chips.forEach((chip) => {
      if (chip.offsetTop > chip.offsetHeight * overflowRowLimit)
        overflowChips++;
    });
  }
  return overflowChips;
};

/**
 * Check if supplied chip object already exists in searchData prop
 * @param {Object} chip - A chip object {lead: 'foo', value: 'bar'}
 * @param {Array} existingArr - An array of chip objects
 */
export const isChipInArray = (chip, existingArr) =>
  existingArr?.some(
    (searchDataItem) =>
      searchDataItem.lead === chip.lead && searchDataItem.value === chip.value,
  );
