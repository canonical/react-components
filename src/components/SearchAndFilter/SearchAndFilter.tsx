import React, { useState, useEffect, useRef } from "react";

import FilterPanelSection from "./FilterPanelSection";
import Chip from "../Chip";
import { overflowingChipsCount, isChipInArray } from "./utils";
import type { SearchAndFilterChip, SearchAndFilterData } from "./types";
import { useOnEscapePressed, useWindowFitment } from "hooks";

export enum Label {
  AddFilter = "Add filter",
  Clear = "Clear input",
  SearchAndFilter = "Search and filter",
}

export type Props = {
  /**
   * A list of chips to initialise inside the input.
   */
  existingSearchData?: SearchAndFilterChip[];
  /**
   * The data for the filter panel.
   */
  filterPanelData: SearchAndFilterData[];
  /**
   * A function that is called when the search data changes.
   */
  returnSearchData: (searchData: SearchAndFilterChip[]) => void;
  /**
   * A function that is called when the filter panel is shown/hidden.
   */
  onPanelToggle?: () => void;
  /**
   * A function that is called when the height of the search container changes.
   */
  onExpandChange?: () => void;
};

/**
 * This component allows searching and filtering of large datasets.
 */
const SearchAndFilter = ({
  existingSearchData = [],
  filterPanelData,
  returnSearchData,
  onPanelToggle = () => {},
  onExpandChange = () => {},
  ...props
}: Props): JSX.Element => {
  const [searchData, setSearchData] = useState(existingSearchData);
  const [searchTerm, setSearchTerm] = useState("");
  const [filterPanelHidden, setFilterPanelHidden] = useState(true);
  const [searchBoxExpanded, setSearchBoxExpanded] = useState(false);
  const [overflowSearchTermCounter, setOverflowSearchTermCounter] = useState(0);
  const [searchContainerActive, setSearchContainerActive] = useState(false);
  const [maxHeight, setMaxHeight] = useState<number>();

  const searchAndFilterRef = useRef(null);
  const searchContainerRef = useRef(null);
  const searchBoxRef = useRef(null);
  const panel = useRef();

  // Call onPanelToggle when the filterPanelHidden state changes
  useEffect(onPanelToggle, [onPanelToggle, filterPanelHidden]);

  // Call onExpandChange when the search box is expanded or collapsed
  useEffect(onExpandChange, [onExpandChange, searchBoxExpanded]);

  // Return searchData to parent component
  useEffect(() => {
    let mounted = true;
    returnSearchData && mounted && returnSearchData(searchData);
    return () => {
      mounted = false;
    };
  }, [searchData, returnSearchData]);

  const searchOnChange = (searchTerm) => {
    setSearchTerm(searchTerm);
  };

  // Hide manual input form field when search container is inactive
  useEffect(() => {
    const searchContainerClickCheck = (e) => {
      const clickInContainer =
        e.target?.closest(".p-search-and-filter") !== null;
      setSearchContainerActive(clickInContainer);
    };

    document.addEventListener("click", searchContainerClickCheck);

    return () => {
      document.removeEventListener("click", searchContainerClickCheck);
    };
  }, [searchContainerActive]);

  const closePanel = () => {
    setFilterPanelHidden(true);
  };
  useOnEscapePressed(() => closePanel());

  // This useEffect sets up listeners so the panel will close if user clicks
  // anywhere else on the page or hits the escape key
  useEffect(() => {
    const mouseDown = (e) => {
      // Check if click is outside of filter panel
      if (!searchAndFilterRef?.current?.contains(e.target)) {
        // If so, close the panel
        closePanel();
      }
    };

    // Add listener on document to capture click events
    document.addEventListener("mousedown", mouseDown);
    // return function to be called when unmounted
    return () => {
      document.removeEventListener("mousedown", mouseDown);
    };
  }, []);

  // Add passed chip to the searchData array
  const toggleSelected = (chip: SearchAndFilterChip) => {
    const currentSelected = [...searchData];
    if (!isChipInArray(chip, currentSelected)) {
      currentSelected.push(chip);
      setSearchData(currentSelected);
      setSearchTerm("");
    } else {
      const updatedCurrentSelected = currentSelected.filter(
        (currentSelectedChip) => {
          return currentSelectedChip.value !== chip.value;
        }
      );
      setSearchData(updatedCurrentSelected);
    }
  };

  // Remove passed chip from the searchData array
  const removeFromSelected = (chip: SearchAndFilterChip) => {
    if (searchData.includes(chip)) {
      const updatedSelected = searchData.filter(
        (searchDataChip) => searchDataChip !== chip
      );
      setSearchData(updatedSelected);
    }
  };

  // When overflow chips are shown, clicking anywhere outside search area
  // or clicking on a chip should hide them again
  useEffect(() => {
    const hideOverflowChips = (e) => {
      if (
        !e.target.closest(".p-search-and-filter") &&
        e.target.className !== "p-icon--close"
      ) {
        setSearchBoxExpanded(false);
      }
    };
    document.addEventListener("click", (e) => {
      hideOverflowChips(e);
    });
    return () => {
      document.removeEventListener("click", (e) => {
        hideOverflowChips(e);
      });
    };
  }, []);

  const handleSubmit = () => {
    if (searchTerm.trim() !== "") {
      toggleSelected({ value: searchTerm, quoteValue: true });
      setSearchTerm("");
    }
  };

  // If the offsetTop is more than double height of a single chip, consider it
  // overflowing
  const updateFlowCount = function () {
    const chips = searchContainerRef?.current?.querySelectorAll(".p-chip");
    const overflowCount = overflowingChipsCount(chips, 1);
    setOverflowSearchTermCounter(overflowCount);
  };

  // Watch for container resize and recalculate overflow count accordingly
  useEffect(() => {
    const resizeObserverSupported = typeof ResizeObserver !== "undefined";
    const wrapper = searchContainerRef.current;
    let wrapperWidthObserver;
    if (resizeObserverSupported && wrapper) {
      wrapperWidthObserver = new ResizeObserver(() => {
        updateFlowCount();
      });
      wrapperWidthObserver.observe(wrapper);
    } else {
      updateFlowCount();
    }
    return () => {
      resizeObserverSupported && wrapperWidthObserver?.disconnect();
    };
  }, [searchData]);

  useWindowFitment(
    panel.current,
    searchAndFilterRef.current,
    (fitsWindow) => setMaxHeight(fitsWindow.fromBottom.spaceBelow - 16),
    0,
    !filterPanelHidden
  );

  // Add search prompt value to search on Enter key
  const searchPromptKeyDown = (e) => {
    if (e.key === "Enter") {
      handleSubmit();
    }
  };

  const searchBox = searchBoxRef.current;
  const searchContainer = searchContainerRef.current;
  if (
    !searchBoxExpanded &&
    searchBox &&
    searchContainer &&
    overflowSearchTermCounter === 0
  ) {
    if (searchBox.offsetTop > searchContainer.offsetHeight) {
      setSearchBoxExpanded(true);
    }
  }

  // If chips or input field contains values, clear 'em out
  const clearAllSearchTerms = () => {
    setSearchTerm("");
  };

  const placeholder = searchData.length
    ? Label.AddFilter
    : Label.SearchAndFilter;

  return (
    <div
      className="p-search-and-filter"
      ref={searchAndFilterRef}
      onClick={() => filterPanelHidden && setFilterPanelHidden(false)}
      {...props}
    >
      <div
        className="p-search-and-filter__search-container"
        aria-expanded={searchBoxExpanded}
        data-active={searchContainerActive || searchData.length === 0}
        data-empty={searchData.length <= 0}
        ref={searchContainerRef}
      >
        {searchTerm !== "" && (
          <button
            aria-label={Label.Clear}
            className="p-search-and-filter__clear"
            onClick={() => clearAllSearchTerms()}
          >
            <i className="p-icon--close" />
          </button>
        )}
        {Object.values(searchData).map((chip) => {
          return (
            <Chip
              lead={chip.lead}
              value={chip.value}
              key={`search-${chip.lead}+${chip.value}`}
              onDismiss={() => removeFromSelected(chip)}
              selected={true}
              quoteValue={chip.quoteValue}
            />
          );
        })}
        <form
          className="p-search-and-filter__box"
          onSubmit={(e) => {
            e.preventDefault();
            handleSubmit();
          }}
          data-overflowing={searchBoxExpanded}
          ref={searchBoxRef}
        >
          <label className="u-off-screen" htmlFor="search-and-filter-input">
            {searchData.length ? Label.AddFilter : Label.SearchAndFilter}
          </label>
          <input
            autoComplete="off"
            className="p-search-and-filter__input"
            id="search-and-filter-input"
            name="search"
            onChange={(e) => searchOnChange(e.target.value)}
            placeholder={placeholder}
            type="search"
            value={searchTerm}
          />
          <button className="u-off-screen" type="submit">
            Search
          </button>
        </form>
        {overflowSearchTermCounter > 0 && (
          <span
            className="p-search-and-filter__selected-count"
            onClick={() => setSearchBoxExpanded(true)}
            onKeyDown={() => setSearchBoxExpanded(true)}
            role="button"
            tabIndex={0}
          >
            +{overflowSearchTermCounter}
          </span>
        )}
      </div>
      {(filterPanelData.length > 0 || searchTerm.length > 0) && (
        <div
          className="p-search-and-filter__panel"
          aria-hidden={filterPanelHidden}
          ref={panel}
          style={{ maxHeight, minHeight: "5rem", overflowX: "auto" }}
        >
          <div>
            {searchTerm.length > 0 && (
              <div
                className="p-search-and-filter__search-prompt"
                onClick={() => handleSubmit()}
                onKeyDown={(e) => searchPromptKeyDown(e)}
                role="button"
                tabIndex={0}
              >
                Search for{" "}
                <span className="p-search-and-filter__search-query">
                  {searchTerm}
                </span>
                ...
              </div>
            )}
            {filterPanelData.map((filterPanelSectionData) => {
              return (
                <FilterPanelSection
                  key={filterPanelSectionData.id}
                  data={filterPanelSectionData}
                  toggleSelected={toggleSelected}
                  searchData={searchData}
                  searchTerm={searchTerm}
                />
              );
            })}
          </div>
        </div>
      )}
    </div>
  );
};

export default SearchAndFilter;
