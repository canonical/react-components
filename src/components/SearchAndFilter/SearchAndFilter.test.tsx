import { render, screen, waitFor } from "@testing-library/react";
import React from "react";

import SearchAndFilter from "./SearchAndFilter";
import { Label } from "./SearchAndFilter";
import userEvent from "@testing-library/user-event";

const sampleData = [
  {
    id: 0,
    heading: "Clouds",
    chips: [{ value: "Google" }, { value: "AWS" }, { value: "Azure" }],
  },
  {
    id: 1,
    heading: "Regions",
    chips: [
      { value: "us-east1" },
      { value: "us-east2" },
      { value: "us-east3" },
    ],
  },
  {
    id: 3,
    heading: "Owner",
    chips: [{ value: "foo" }, { value: "bar" }, { value: "baz" }],
  },
];
// eslint-disable-next-line testing-library/no-node-access
const getPanel = () => document.querySelector(".p-search-and-filter__panel");

const getSearchContainer = () =>
  // eslint-disable-next-line testing-library/no-node-access
  document.querySelector(".p-search-and-filter__search-container");

describe("Search and filter", () => {
  it("renders", async () => {
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        data-testid="searchandfilter"
        filterPanelData={[]}
        returnSearchData={returnSearchData}
      />
    );
    expect(screen.getByTestId("searchandfilter")).toMatchSnapshot();
  });

  it("hide the clear button when there is no value in search box", async () => {
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={[]}
        returnSearchData={returnSearchData}
      />
    );
    expect(
      screen.queryByRole("button", { name: Label.Clear })
    ).not.toBeInTheDocument();
  });

  it("shows panel when clicking the wrapping element", async () => {
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        data-testid="searchandfilter"
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
      />
    );
    expect(getPanel()).toHaveAttribute("aria-hidden", "true");
    await waitFor(async () => {
      await userEvent.click(screen.getByTestId("searchandfilter"));
    });
    expect(getPanel()).toHaveAttribute("aria-hidden", "false");
  });

  it("shows panel on click", async () => {
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
      />
    );
    expect(getPanel()).toHaveAttribute("aria-hidden", "true");
    await waitFor(async () => {
      await userEvent.click(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    expect(getPanel()).toHaveAttribute("aria-hidden", "false");
  });

  it("should hide chip overflow counter when none overflow", async () => {
    const returnSearchData = jest.fn();
    // Jest is unaware of layout so we must mock the offsetTop and offsetHeight
    // of the chips
    Object.defineProperty(HTMLElement.prototype, "offsetHeight", {
      configurable: true,
      value: 40,
    });
    Object.defineProperty(HTMLElement.prototype, "offsetTop", {
      configurable: true,
      value: 40,
    });
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
      />
    );
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-search-and-filter__selected-count")
    ).not.toBeInTheDocument();
  });

  it("show overflow chip counter when chips overflow", async () => {
    // Jest is unaware of layout so we must mock the offsetTop and offsetHeight
    // of the chips
    Object.defineProperty(HTMLElement.prototype, "offsetHeight", {
      configurable: true,
      value: 40,
    });
    Object.defineProperty(HTMLElement.prototype, "offsetTop", {
      configurable: true,
      value: 100,
    });
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
      />
    );
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-search-and-filter__selected-count")
    ).not.toBeInTheDocument();
    await waitFor(async () => {
      await userEvent.click(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    await waitFor(async () => {
      await userEvent.click(screen.getByRole("button", { name: "us-east1" }));
    });
    expect(screen.getByRole("button", { name: "+1" })).toBeInTheDocument();
  });

  it("all chips are shown when counter is clicked", async () => {
    // Jest is unaware of layout so we must mock the offsetTop and offsetHeight
    // of the chips
    Object.defineProperty(HTMLElement.prototype, "offsetHeight", {
      configurable: true,
      value: 40,
    });
    Object.defineProperty(HTMLElement.prototype, "offsetTop", {
      configurable: true,
      value: 100,
    });
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
      />
    );
    expect(getSearchContainer()).toHaveAttribute("aria-expanded", "false");
    await waitFor(async () => {
      await userEvent.click(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    await waitFor(async () => {
      await userEvent.click(screen.getByRole("button", { name: "us-east1" }));
    });
    await waitFor(async () => {
      await userEvent.click(screen.getByRole("button", { name: "+1" }));
    });
    expect(getSearchContainer()).toHaveAttribute("aria-expanded", "true");
  });

  it("search prompt appears when search field has search term", async () => {
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
      />
    );
    await waitFor(async () => {
      await userEvent.click(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-search-and-filter__search-prompt")
    ).not.toBeInTheDocument();
    await waitFor(async () => {
      await userEvent.clear(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    await waitFor(async () => {
      await userEvent.type(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter }),
        "My new value"
      );
    });
    expect(
      screen.getByRole("button", { name: "Search for My new value ..." })
    ).toBeInTheDocument();
  });

  it("no search results appear for unknown search term", async () => {
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
      />
    );
    await waitFor(async () => {
      await userEvent.click(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelectorAll(".p-filter-panel-section").length).toEqual(
      3
    );
    await waitFor(async () => {
      await userEvent.clear(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    await waitFor(async () => {
      await userEvent.type(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter }),
        "Unknown value"
      );
    });
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelectorAll(".p-filter-panel-section").length).toEqual(
      0
    );
  });

  it("correct number of panels appear for matching search terms", async () => {
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
      />
    );
    await waitFor(async () => {
      await userEvent.click(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    await waitFor(async () => {
      await userEvent.clear(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    await waitFor(async () => {
      await userEvent.type(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter }),
        "Google"
      );
    });
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelectorAll(".p-filter-panel-section").length).toEqual(
      1
    );
    await waitFor(async () => {
      await userEvent.clear(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    await waitFor(async () => {
      await userEvent.type(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter }),
        "re"
      );
    });
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelectorAll(".p-filter-panel-section").length).toEqual(
      2
    );
  });

  it("Matching search terms are highlighted with strong tag", async () => {
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
      />
    );
    await waitFor(async () => {
      await userEvent.click(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    await waitFor(async () => {
      await userEvent.clear(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    await waitFor(async () => {
      await userEvent.type(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter }),
        "Google"
      );
    });
    const boldText = document
      // eslint-disable-next-line testing-library/no-node-access
      .querySelectorAll(".p-chip")[0]
      // eslint-disable-next-line testing-library/no-node-access
      .querySelector("strong");
    expect(boldText?.textContent).toEqual("Google");
  });

  it("Existing search data displays correctly", async () => {
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
        existingSearchData={[{ lead: "Cloud", value: "Google" }]}
      />
    );
    // eslint-disable-next-line testing-library/no-node-access
    const lead = document.querySelector(
      ".p-search-and-filter__search-container .p-chip__lead"
    )?.textContent;
    // eslint-disable-next-line testing-library/no-node-access
    const value = document.querySelector(
      ".p-search-and-filter__search-container .p-chip__value"
    )?.textContent;
    expect(lead).toBe("CLOUD");
    expect(value).toBe("Google");
  });

  it("Existing search data displays correctly if two search items present", async () => {
    const returnSearchData = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
        existingSearchData={[
          { lead: "Cloud", value: "Google" },
          { lead: "Region", value: "eu-west-1" },
        ]}
      />
    );
    // eslint-disable-next-line testing-library/no-node-access
    const chips = document.querySelectorAll(
      ".p-search-and-filter__search-container .p-chip"
    );
    expect(chips.length).toBe(2);

    // eslint-disable-next-line testing-library/no-node-access
    const chip1Value = document.querySelector(
      ".p-search-and-filter__search-container .p-chip:nth-child(1) .p-chip__value"
    )?.textContent;
    expect(chip1Value).toEqual("Google");

    // eslint-disable-next-line testing-library/no-node-access
    const chip2Value = document.querySelector(
      ".p-search-and-filter__search-container .p-chip:nth-child(2) .p-chip__value"
    )?.textContent;
    expect(chip2Value).toEqual("eu-west-1");
  });

  it("calls onHeightChange function when provided and the counter is clicked", async () => {
    // Jest is unaware of layout so we must mock the offsetTop and offsetHeight
    // of the chips
    Object.defineProperty(HTMLElement.prototype, "offsetHeight", {
      configurable: true,
      value: 40,
    });
    Object.defineProperty(HTMLElement.prototype, "offsetTop", {
      configurable: true,
      value: 100,
    });
    const returnSearchData = jest.fn();
    const onHeightChange = jest.fn();
    render(
      <SearchAndFilter
        filterPanelData={sampleData}
        returnSearchData={returnSearchData}
        onHeightChange={onHeightChange}
      />
    );
    await waitFor(async () => {
      await userEvent.click(
        screen.getByRole("searchbox", { name: Label.SearchAndFilter })
      );
    });
    await waitFor(async () => {
      await userEvent.click(screen.getByRole("button", { name: "us-east1" }));
    });
    await waitFor(async () => {
      await userEvent.click(screen.getByRole("button", { name: "+1" }));
    });
    expect(onHeightChange).toHaveBeenCalled();
  });
});
