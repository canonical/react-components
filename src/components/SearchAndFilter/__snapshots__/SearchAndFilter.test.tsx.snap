// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search and filter renders 1`] = `
<div
  class="p-search-and-filter"
  data-testid="searchandfilter"
>
  <div
    aria-expanded="false"
    class="p-search-and-filter__search-container"
    data-active="true"
    data-empty="true"
  >
    <form
      class="p-search-and-filter__box"
      data-overflowing="false"
    >
      <label
        class="u-off-screen"
        for="search-and-filter-input"
      >
        Search and filter
      </label>
      <input
        autocomplete="off"
        class="p-search-and-filter__input"
        id="search-and-filter-input"
        name="search"
        placeholder="Search and filter"
        type="search"
        value=""
      />
      <button
        class="u-off-screen"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
</div>
`;
